arangod/Aql/RestAqlHandler.cpp

```diff
@@ -243,9 +247,35 @@ void RestAqlHandler::setupClusterQuery() {
 
   answerBuilder.close(); // result
   answerBuilder.close();
-  
+
+  QueryId queryId = q->id();   // keep a copy
   _queryRegistry->insertQuery(std::move(q), ttl);
 
+  // Now set an alarm for the case that the coordinator is restarted which
+  // initiated this query. In that case, we want to drop our piece here:
+  bool found;
+  std::string const& coordinatorId = _request->header(StaticStrings::ClusterCommSource, found);
+  if (found) {
+    // If we do not see this header, then we do not know which coordinator issues
+    // this request, then we simply do nothing and let the ttl be the only safety net.
+    auto& clusterFeature = _server.getFeature<ClusterFeature>();
+    auto& clusterInfo = clusterFeature.clusterInfo();
+    RebootId rebootId = clusterInfo.getCurrentRebootId(coordinatorId);
+    if (rebootId.initialized()) {
```

Comment:
Still wondering why we don't create a CallbackGuard object before we call QueryRegistry::insertQuery(...), and then pass this object into the method.
That will avoid the duplicate lookup in `_queries` and also make the insert operation atomic.

Reply:
> `/_api/aql/setup` is an API - admittedly an internal one, and must be prone against bad input values. If the `coordinatorRebootIdSlice` is represented as a small int or int VPack then `.getUInt()` might throw if a negative value is presented. Since such RebootIds are nonsensical, we simply do nothing in this case w.r.t. the reboot tracker. However, I believe that we should guard against this exception.

I disagree with that. I do not think an API must always do something, even if it's used wrongly. That just makes things go silently wrong instead of aborting safely. In this case, say a client passes the reboot ID in a wrong way. It will be silently ignored, and the author has no chance to fix his or her error. But the protection you're carefully implementing here, and is assumed at the client to work, will just be silently annulled, and no one will notice until it goes wrong.

> Still wondering why we don't create a CallbackGuard object before we call QueryRegistry::insertQuery(...), and then pass this object into the method.
That will avoid the duplicate lookup in _queries and also make the insert operation atomic.

Makes sense to me. Though it is unrelated to the `initialized` check, or am I missing something?